[docs] add hello-dagster (#10684)
dagster-io/dagster: docs/content/getting-started/hello-dagster.mdx

    """
    top_story_ids = requests.get(
        "https://hacker-news.firebaseio.com/v0/topstories.json"
    ).json()
    return top_story_ids[:10]

# asset dependencies can be inferred from inputs
@asset
def hackernews_top_stories(hackernews_top_story_ids):
    """Get items based on story ids from the HackerNews items endpoint"""

    results = []
    for item_id in hackernews_top_story_ids:
        item = requests.get(
            f"https://hacker-news.firebaseio.com/v0/item/{item_id}.json"
        ).json()
        results.append(item)

    df = pd.DataFrame(results)

    # recorded metadata can be customized
    metadata = {
        "num_records": len(df),
        "preview": MetadataValue.md(df.to_markdown()),
    }

    return Output(value=df, metadata=metadata)

@repository
def hello_dagster():
    return [hackernews_top_story_ids, hackernews_top_stories]
```

Install the Python packages you'll need to run `hello-dagster.py` in your favorite Python environment. Unsure? Follow our [installation guide](/getting-started/install).

```bash
# run in a terminal in your favorite python environment
pip install dagster dagit pandas
```

> Next, run`dagit` in your terminal in the same directory as `hello-dagster.py`. This command starts Dagster's user interface.

Author: ```suggestion
Next, run `dagit` in your terminal in the same directory as `hello-dagster.py`. This command starts Dagster's user interface.
```

---
title: Hello Dagster | Dagster
description: Run dagster for the first time
---

# Hello Dagster

> This guide helps you build a simple data pipeline in dagster that downloads the top 10 HackerNews stories.

Reviewer: I like that the intro tells me what I can expect to have by the end, but this doesn't draw me in. How long will this take me? Why should I do it?

Author: ```suggestion
This guide helps you build a simple data pipeline in dagster that downloads the top 10 HackerNews stories. In less than 5 minutes you will be able to develop functional code and explore dagster's user interface.
```

> Start by creating a file called `hello-dagster.py` and copying the following:

Reviewer: Happy to do it!

import pandas as pd
import requests

from dagster import MetadataValue, Output, asset, repository

@asset
def hackernews_top_story_ids():
    """
    Get top stories from the HackerNews top stories endpoint.
    API Docs: https://github.com/HackerNews/API#new-top-and-best-stories
    """
    top_story_ids = requests.get(
        "https://hacker-news.firebaseio.com/v0/topstories.json"
    ).json()
    return top_story_ids[:10]

# asset dependencies can be inferred from inputs
@asset
def hackernews_top_stories(hackernews_top_story_ids):
    """Get items based on story ids from the HackerNews items endpoint"""

    results = []
    for item_id in hackernews_top_story_ids:
        item = requests.get(
            f"https://hacker-news.firebaseio.com/v0/item/{item_id}.json"
        ).json()
        results.append(item)

    df = pd.DataFrame(results)

    # recorded metadata can be customized
    metadata = {
        "num_records": len(df),
        "preview": MetadataValue.md(df.to_markdown()),
    }

    return Output(value=df, metadata=metadata)

@repository
def hello_dagster():
    return [hackernews_top_story_ids, hackernews_top_stories]
```

> Now install the Python packages we'll need to run `hello-dagster.py` in your favorite Python environment. Unsure? Follow our [installation guide](/getting-started/install).

Reviewer: ```suggestion
Install the Python packages you'll need to run `hello-dagster.py` in your favorite Python environment. Unsure? Follow our [installation guide](/getting-started/install).
```

Suggesting we change this from `we` to `you`. It read a little bit odd to me.

```bash
# run in a terminal in your favorite python environment
pip install dagster dagit pandas
```

> We'll use `dagit`, the dagster user interface, to run `hello-dagster.py`:

Reviewer: A few things here:

```suggestion
Using Dagit, the Dagster user interface, run `hello-dagster.py`:
```

- Dagit and Dagster should be capitalized. I think here you were referring to the command to start Dagit, but the sentence reads like it's actually the name / proper noun.

   You could say something like "Next, run `dagit` in your terminal to start Dagit, Dagster's user interface" and that would work.
- The instructions should clarify that the user needs to run this command in the same location as `hello-dagster.py`. I think it's implied, but given the number of issues beginners have I think being overly specific is fine / preferred

> We'll use `dagit`, the dagster user interface, to run `hello-dagster.py`:

Author: ```suggestion
Next, run`dagit` in your terminal in the same directory as `hello-dagster.py`. This command starts Dagster's user interface.
```

```bash
# run in a terminal in your favorite python environment
dagit -f hello-dagster.py
```

> Navigate your browser to `http://localhost:3000/`. Clicking `Materialize All` will run our pipeline and create our assets.

Reviewer: ```suggestion
In your browser, navigate to [http://localhost:3000/](http://localhost:3000). Click **Materialize All** to run the pipeline and create your assets.
```

Suggesting that `localhost` be a link, so users can just click it if they want to. Also changed the formatting for the button text to bold to be consistent with the style changes I'm making (but aren't necessarily in the style guide yet, sorry 😬 )

<Image
alt="HackerNews asset graph"
src="/images/getting-started/hello-dagster/hello-dagster.png"
width={1331}
height={874}
/>

> The `hackernews_top_stories` asset included `metadata`. Click on the asset and then `Show Markdown` to preview the pipeline result and view the top HackerNews stories.

Reviewer: This is a really cool part of the example that feels sorta odd in its current state. Perhaps something like this?:

```suggestion
That's it! You now have two materialized Dagster assets. 

But wait - there's more. Because the `hackernews_top_stories` asset specified `metadata`, you can view the metadata right in Dagit.

Click the asset. In the sidebar that displays, click the **Show Markdown** link in the **Materialization in Last Run** section. This opens a preview of the pipeline result, allowing you to view the top 10 HackerNews stories:
```

I think it'd be cool to show an image of the preview here as well, or a GIF that goes from clicking **Show Markdown** to the preview opening. It admittedly took me a moment to locate the link in the UI, and I'd like to reduce the amount of thinking folks have to do.

> ## Next Steps

Reviewer: ```suggestion
---

## Next steps
```

Headings should use sentence case, and H2s should be preceded by a horizontal rule. It helps break things up visually

> This example uses [assets](/tutorial#assets). Most new Dagster projects use assets because they let data engineers:

Reviewer: What's the thinking for linking to the tutorial over a different page?

Author: No strong opinion on the specific next step link. I think once the new hackernews full fledged example / tutorial is done that would be a great next step that builds off of the sample code

Reviewer: Definitely agreed there. Given that we're hoping to release that soon, I think this is okay to leave as-is

> - think in the same terms as stakeholders
> - answer questions about data quality and lineage
> - work with the modern data stack (dbt, Airbyte/Fivetran, Spark)
> - create declarative freshness policies instead of task-driven cron schedules

Reviewer: ```suggestion
- Think in the same terms as stakeholders
- Answer questions about data quality and lineage
- Work with the modern data stack (dbt, Airbyte/Fivetran, Spark)
- Create declarative freshness policies instead of task-driven cron schedules
```

> Dagster also offers [ops and jobs](/tutorial#ops-and-jobs), but we recommend starting with assets.

Reviewer: Is the expectation that users should know what these things are? This personally feels out of place, but that might just be me

Author: My goal was to clarify that most Dagster projects should start with assets, since a very common question I receive is "should I use assets or ops?"

Reviewer: Yep; I get that. I was thinking about this as more of "what are ops and jobs and why should I care?" but perhaps users coming in will already have some idea what these are?

> Our `hello-dagster` example used a single file, but most Dagster projects are organized as Python packages. Check out our [getting started guide](/getting-started/create-new-project) to begin with a scaffolded blank project or start from an official example such as this [dagster + dbt project](/integrations/dbt/using-dbt-with-dagster).

Reviewer: ```suggestion
While this example used a single file, most Dagster projects are organized as Python packages. From here, you can:

- Start with a scaffolded blank project. Check out the [new project guide](/getting-started/create-new-project) for more info.
- Start with an official example, such as the [dbt + Dagster project](/integrations/dbt/using-dbt-with-dagster). Check out [all the examples in GitHub](https://github.com/dagster-io/dagster/tree/master/examples).
```

Suggesting we break this up a bit and put the choices into a list to make them easier to parse. I also added a link to the examples repo in GH in case folks want to peruse.

---
title: Hello Dagster | Dagster
description: Run dagster for the first time
---

# `hello dagster`

This guide helps you build a simple data pipeline in dagster that downloads the top 10 HackerNews stories.

1. Start by creating a file called `hello-dagster.py` and copying the following:

    ```python file=/getting-started/hello-dagster/hello-dagster.py

>     # copy to a file called hello-dagster.py

Reviewer 2: Can we take this comment out?

    from dagster import repository, asset, MetadataValue
    import requests
    import pandas as pd

>     # dagster helps you think about datasets, models, 

Reviewer 2: Can we take this comment out?

    # and other objects that you want to exist as assets
    # instead of worrying about tasks

    @asset
    def hackernews_topstory_ids():
        """
        Get up to 500 top stories from the HackerNews topstories endpoint.
        API Docs: https://github.com/HackerNews/API#new-top-and-best-stories
        """

>         newstories_url = "https://hacker-news.firebaseio.com/v0/topstories.json"

Reviewer 2: Should `newstories_url` be `topstories_url`? Or inlined with the function call? And `top_500_newstories` be `top_500_stories`? I found myself trying to figure out what a "newstory" was.

        top_500_newstories = requests.get(newstories_url).json()
        return top_500_newstories

>     # dependencies between assets can be expressed as function arguments

Reviewer 2: Can we take this comment out? I think it's infer-able from looking at the code.

Reviewer 3: I actually think it's hard to understand what's happening. I think inferring the upstream asset from the name of the arg is fairly magical and we need to include an explanation somewhere, whether it's as as comment, or commentary after the code sample.

Author: I'll add back a comment which I think strikes a nice balance between Sandy's concerns for code pars-ability while still calling out the magical pieces.

    @asset 
    def hackernews_topstories(context, hackernews_topstory_ids):

>         """

Reviewer 2: We could make this more concise by putting the docstring on one line and removing the newline after

        Get items based on story ids from the HackerNews items endpoint
        """

        results = []
        for item_id in hackernews_topstory_ids:
            item = requests.get(f"https://hacker-news.firebaseio.com/v0/item/{item_id}.json").json()
            results.append(item)

>             if len(results) >= 10:

Reviewer 2: I would expect `hackernews_topstories` to contain the same stories as `hackernews_topstory_ids`. Thoughts on applying this limit in `hackernews_topstory_ids`?  It could be more concise there - just `return top_500_newstories[:min(len(top_500_newstories), 10)]`

                break

        df = pd.DataFrame(results)

        # dagster tracks rich metadata about our assets

>         context.add_output_metadata(

Reviewer 2: We typically recommend adding output metadata using 
```
return Output(
    value=df,
    metadata={...},
)
```

Reviewer 3: is this documented anywhere? i have seen this pattern before but i don't think it's in our api docs anywhere

Reviewer 3: also - i wonder if we should just skip including metadata as part of the example. seems like 2 extra concepts (what's `Output`? what's `metadata`?) that we may want to introduce more gradually

Author: The documentation on output metadata is definitely in need of some love. I included it in this example so that when someone runs the pipeline they can see the hackernews data somewhere. Otherwise it is tricky to validate that dagster did anything without determining how to load the asset from a pickle file in an ugly tmp directory.

            {
                "num_records": len(df),
                "preview": MetadataValue.md(df.to_markdown()),
            }
        )
        return df

    @repository
    def hello_dagster():
        return [hackernews_topstory_ids, hackernews_topstories]
    ```

2. Now install the Python packages we'll need to run `hello-dagster.py` in your favorite Python environment. Unsure? Follow our [installation guide](/getting-started/install).

    ```bash
    # run in a terminal in your favorite python environment
    pip install dagster dagit pandas
    ```

3. We'll use `dagit`, the dagster user interface, to run `hello-dagster.py`:

    ```bash
    # run in a terminal in your favorite python environment
    dagit -f hello-dagster.py
    ```

Navigate your browser to `http://localhost:3000/`. Clicking `Materialize All` will run our pipeline and create our assets.

<Image
alt="HackerNews asset graph"
src="/images/getting-started/hello-dagster/hello-dagster.png"
width={1331}
height={874}
/>

## Next Steps

> This example uses [assets](/tutorial#assets). Most new Dagster projects use assets because they let us:

Reviewer 2: We typically avoid using the first person in this way, because it gets confusing whether "we/us" means "the Dagster team" (like how you're using it in "we recommend starting with assets") vs. "˝data practitioners" (like how you're using it in "Most new Dagster projects use assets because they let us".
